### ISSUE TITLE
Recipe: Cache CSS and fonts

### ISSUE THREAD

**Issue by digitarald** [mozilla/serviceworker-cookbook#9]
1. Serve stale files for CSS and fonts on load
2. Fetch new version into cache (on install or on fetch?)
  - Fonts probably didn't update but just CSS, should the service worker only invalidate changed files?
3. Refresh CSS in client when updates are available

Similar to https://github.com/phamann/embrace-the-network/tree/master/src/stale-while-revalidate

**Comment by digitarald** [comment]
Taking a shot at the caching strategy here …

*CSS assets* = CSS and media referenced from CSS with low rate of updates (fonts, icons and spritesheets)

- Cold cache
  1. SW is installed
  2. CSS assets added to cache in `onfetch`
- Warm cache
  1. CSS assets served from cache via `onfetch`
  2. Check if a an update request is needed, could be
    - JSON file that references files and *hashes*/*last-modified*
    - Cache age < configured time (say 6h)
 3. If cache needs an updated
    1. Cache updated fetched assets
    2. Notify frontend of update CSS
    3. Client refreshes CSS files

@lodr what do you think is the best strategy here?

**Comment by delapuente** [comment]
I was thinking about this. I agree on cold-cache strategy. Assuming [Request](https://fetch.spec.whatwg.org/#requests) is completely implemented, we can check `type` and `destination` along with `content-type` to identify fonts and CSS.

Fir the first version, the warm cache strategy @digitarald is suggesting coudl work but the problem I see is that an update in a CSS file could imply an update in the dependant HTML. This is something we should look after. Once determined if the new CSS / font is requires, cahing and sending a signal to the client to perform a reload in a controlled way seems to be the best.

**Comment by digitarald** [comment]
> Assuming Request is completely implemented, we can check type and destination along with content-type to identify fonts and CSS.

Request is complete aside from `cache`. Checking content-type would a good way to determine CSS

> Fir the first version, the warm cache strategy @digitarald is suggesting coudl work but the problem I see is that an update in a CSS file could imply an update in the dependant HTML. This is something we should look after.

CSS updates are less critical though and the flash-of-mismatching-css/html should be only a short one as the `sw` should trigger an update within a few hundred ms after page load. Does that make sense?

**Comment by delapuente** [comment]
Yes, it should be short and ultimately, the SW don't cause the reload, it just notifies the client and it's the client which is in charge of reloading the page. What we should put in the recipe is how to reload CSS only (without realoading the entire HTML) by adding a bumping query-parameter that the service worker ignores when looking for the new CSS.

Deciding when to update is more complicated. How did you plan to do? With a _cache-age strategy_, we should annotate last cache time somewhere (in a custom header inside the response, for instance) then use it to compare with present time to see if another request is needed.

The problem with a JSON strategy is you need to download the JSON file each time to see if something changed.

An alternative combining hashes and cache-age is as follows:
  0. When cold-caching, annotate caching time in the response as a header.
  1. When warm-caching, retrieve the resource and read the caching time and ETag.
  2. Compare with now. If max cache-age is surpased, perform a HEAD request and get the ETag.
  3. Compare with former ETag, if they differ, perform the complete FETCH on the resource and recache.

What do you think @digitarald ?

**Comment by TimAbraldes** [comment]
Hey guys, good discussion going on here!

I implemented a first pass at this in #63 that I think is pretty informative as a recipe. I wanted to avoid needing any server-side logic and tried to make the example as self-explanatory as possible. Please try running it in Chrome and let me know if I succeeded. ISTM that a lot of the points brought up in the discussion here would be totally relevant to production code but maybe could be glossed over in the recipe?

Let me know what you think!

**Comment by delapuente** [comment]
When reviewing the PR from @TimAbraldes , I was supponsing some use case but I think I'm not getting the point of this recipe. @digitarald do you mind to explain it to me a little bit more? What is the use case and what is the solution overview? Thanks!
